freezed_provider_value_notifier/lib/todo_list_model.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 The Flutter Architecture Sample Authors. All rights reserved.
+// Use of this source code is governed by the MIT license that can be found
+// in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/foundation.dart';
+import 'package:collection/collection.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:freezed_provider_value_notifier/models.dart';
+import 'package:todos_repository_core/todos_repository_core.dart';
+
+part 'todo_list_model.freezed.dart';
+
+enum VisibilityFilter { all, active, completed }
+
+@freezed
+abstract class TodoList with _$TodoList {
+  factory TodoList(
+    List<Todo> todos, {
+    @required VisibilityFilter filter,
+    @required bool loading,
+  }) = TodoListState;
+
+  @late
+  int get numCompleted =>
+      todos.where((Todo todo) => todo.complete).toList().length;
+
+  @late
+  bool get hasCompleted => numCompleted > 0;
+
+  @late
+  int get numActive =>
+      todos.where((Todo todo) => !todo.complete).toList().length;
+
+  @late
+  bool get hasActiveTodos => numActive > 0;
+
+  @late
+  List<Todo> get filteredTodos => todos.where((todo) {
+        switch (filter) {
+          case VisibilityFilter.active:
+            return !todo.complete;
+          case VisibilityFilter.completed:
+            return todo.complete;
+          case VisibilityFilter.all:
+          default:
+            return true;
+        }
+      }).toList();
+}
+
+extension TodoById on TodoList {
+  Todo todoById(String id) =>
+      todos.firstWhere((it) => it.id == id, orElse: () => null);
+}
+
+class TodoListController extends ValueNotifier<TodoList> {
+  TodoListController({
+    VisibilityFilter filter = VisibilityFilter.all,
+    @required this.todosRepository,
+    List<Todo> todos = const [],
+  })  : assert(todosRepository != null),
+        super(TodoList(todos, filter: filter, loading: false)) {
+    _loadTodos();
+  }
+
+  final TodosRepository todosRepository;
+
+  set filter(VisibilityFilter filter) =>
+      setState(value.copyWith(filter: filter));
+
+  void setState(TodoList state) {
```

Comment:
Would it make sense to call this something like `updateValue` instead of `setState`? I was a bit confused when I first read this code thinking it might be tied to a Widget.

Reply:
I removed it in value of `value = newValue`.

Reply:
I dig it. Adding `protected` also makes me happy, since I did notice the `value` could have been changed by anyone at any time from an outside source.

Sorry, moved from another thread -- not sure how this comment got attached to the other one :/

Reply:
Wait -- this now triggers an infinite loop!

Reply:
Oopsy~